aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -3516,14 +3516,16 @@ static void linalg_lstsq_out_info(
       at::sum_out(residuals, raw_residuals, /*dim=*/-2, /*keepdim=*/false, /*dtype*/real_dtype);
     }
   }
-  solution = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
+  auto solution_view = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
+  // manually restride original
+  solution.set_(solution.storage(), solution_view.storage_offset(), solution_view.sizes(), solution_view.strides());
```

Comment:
What has changed with `narrow` and why is this change needed?

Reply:
@ezyang can comment on this more tomorrow, but to quote: "writing over Tensor& is not kosher".

This was causing some double-free errors elsewhere.

Reply:
Yes, writing over Tensor&, in general, should be avoided.
Why does it have to be in this PR?

Reply:
Enabling `enable_python_dispatcher` in `test/test_decomp.py` triggered errors with this code and corrupted memory.

<img width="855" alt="image" src="https://user-images.githubusercontent.com/6355099/191179365-2c5110fc-f329-4fc9-87e9-5dbbb8575325.png">

https://github.com/pytorch/pytorch/actions/runs/3085755988/jobs/4989537077

Reply:
Why do you peek inside the `CompositeExplicitAutograd` function at all?
Maybe correct fix would be to change `CompositeExplicitAutograd: linalg_lstsq -> CPU, CUDA: linalg_lstsq` here: https://github.com/pytorch/pytorch/blob/bcdef58a55665bba959a773633aa5b1e5758d94e/aten/src/ATen/native/native_functions.yaml#L12246 ?

Reply:
We're not peeking inside `CompositeExplicitAutograd`. `enable_python_dispatcher` allows us to register python implementations of `CompositeImplicitAutograd` functions (like you've been asking for!), and this PR adds a python implementation of `aten::matmul`. `linalg_lstsq` backward formula uses a matmul, and so we enter the decomposition for it (where we see the shape error).

Reply:
> like you've been asking for

I don't remember asking for it 😅 

There was no shape error before this PR, could it mean that Python's implementation in this PR is different from `aten::matmul`?

Reply:
Haha, I just mean that this allows us to make use of `CompositeImplicitAutograd` decompositions, while previously they weren't so useful :P 

> could it mean that Python's implementation in this PR is different from aten::matmul

Yeah, that's a reasonable thing to guess. But in this case, the test was failing from `enable_python_dispatcher` even without adding the `matmul` decomp :P

Reply:
It's a memory error, caused by the overwriting of the mutable Tensor&. Once you have a memory error, all bets are off; in this case, it manifests as the shape mismatched, but the true problem is the memory error. You can determine this by running the program with ASAN with the python dispatcher enabled.

The reason the python dispatcher matters is because it causes all dispatched calls (that includes the internal `at::squeeze`, etc. calls inside the CompositeExplicitAutograd) to go into the boxed calling convention. In fact, it is not even necessary to have Python at all; forcing all dispatcher calls through the boxed calling convention is sufficient to trigger the ASAN error. I am guessing what is happening is that when we put the Tensors onto the stack, we avoid refcount bump (because the invariant is that input tensors will stay live for the entirety of the function). When you edit the mutable reference, this invariant is broken (the tensor is freed); ordinarily this is fine because the reference also got updated, but if there is a copy of the Tensor ref on the stack, that reference doesn't get updated, and thus you have the problem.

If you like, we can separate out the "avoid Tensor& modification" into its own patch. But the new code is clearly more correct than the old code.

Reply:
Yes, I would like it to be separated into its own patch so that changes to the torch.linalg part of the project are clear and well-tracked.